### PR TITLE
[gilbert] Tangent-frame wall-shear loss targeting wsy/wsz gap (Wave 20)

### PR DESCRIPTION
**Pull request by morganmcg1**
# [gilbert] Tangent-frame wall-shear loss (Wave 20)

## Hypothesis

**The wsy/wsz error is dominated by orientation, not magnitude.** Wall shear is a vector tangent to the surface — predicting it in **world (x, y, z) coordinates** forces the model to rotate the same physical quantity into a global frame that depends on vehicle pose. Predicting it in the **local surface tangent basis** (t1, t2, n) would let the model output a 2D in-plane vector aligned with the surface, removing the dominant source of rotational variance and dramatically reducing wsy/wsz error.

The current best (PR #174, 5L/256d) gives:
- wsx = 7.109% (target 5.35%, gap 1.76pp)
- **wsy = 8.7345%** (target 3.65%, gap **5.08pp**)
- **wsz = 10.5766%** (target 3.63%, gap **6.95pp**)

The y/z components dominate the abupt residual. AB-UPT achieves 3.65/3.63% on these — strong evidence that an architectural representation change (not just a loss reweighting) is the missing piece. Tangent-frame prediction is the standard physics-motivated formulation for surface vector fields and has not been tried on this codebase.

## Implementation

Add a tangent-frame option that:
1. **Compute per-point tangent basis from surface normals.** For each surface point, build a local orthonormal frame (t1, t2, n) where n = surface normal, t1 = unit vector orthogonal to n along the dominant principal direction (or any deterministic in-plane construction — Gram-Schmidt against world-x is fine), t2 = n × t1. The basis must be deterministic so train/val use the same transform.
2. **Rotate target wall-shear into local frame.** For each surface point with world-frame target shear `[ws_x, ws_y, ws_z]`, project to local: `ws_local = R^T · ws_world` where R = [t1 | t2 | n]. Since wall shear is tangent to the surface, the n-component should be near zero by construction; you can either (a) keep all 3 components and let the model learn that ws_n ≈ 0, or (b) zero out ws_n explicitly and only supervise on (t1, t2) — option (b) is cleaner.
3. **Model still predicts 3 channels per point** in local frame. At inference, rotate back: `ws_world_pred = R · ws_local_pred`.
4. **Compute val metrics in world frame** (i.e. after rotation back) so abupt_axis_mean_rel_l2_pct stays comparable to the baseline and to AB-UPT.

Add a `--tangent-frame-shear` flag (default `False`). When enabled, do the loss in local frame; when disabled, behavior is identical to PR #174 baseline.

**You will need surface normals.** Check `target/data` and the manifest — DrivAerML provides per-point normals on the surface. If normals are not in the loaded tensors, add them to the dataloader. If normals are unavailable (only positions), compute via local-PCA on k-nearest-neighbours (k=8) at preprocessing time and cache.

## Trial plan

**Trial A — minimal smoke test (8 epochs, sub-sample dataset OK if needed for debugging):**
- Confirm forward+backward works, val rotates back correctly, train loss is sane (similar magnitude to baseline within 2x).
- Confirm wsy/wsz val numbers at ep5 are not catastrophically worse than baseline at ep5 (~11–12%). If ep5 wsy/wsz > 18%, kill — implementation bug.

**Trial B — full 50-epoch run** (only if Trial A passes ep5 sanity):
- 5L/256d, FourierPE, no-EMA, lr=3e-4, T_max=50, batch-size=2, **--tangent-frame-shear**.
- Compare to baseline PR #174 ep~45 (val_abupt=6.9549%, wsy=8.73%, wsz=10.58%).
- Target: wsy < 7.5%, wsz < 9.0% (≥1pp drop on each).

## Reproduce command (Trial B, run once Trial A passes)

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --fourier-pe --no-use-ema \
  --lr 3e-4 --lr-cosine-t-max 50 \
  --no-compile-model \
  --tangent-frame-shear \
  --kill-thresholds "500:train/loss<5,89000:val_primary/abupt_axis_mean_rel_l2_pct<13,178000:val_primary/abupt_axis_mean_rel_l2_pct<10.5,267000:val_primary/abupt_axis_mean_rel_l2_pct<8.5" \
  --wandb-group bengio-wave20 --wandb-name gilbert/tangent-frame-shear --agent gilbert
```

## Kill gates (standard)
- EP5 (step 89,080) abupt > 13.0% → kill
- EP10 (step 178,160) abupt > 10.5% → kill
- EP15 (step 267,240) abupt > 8.5% → kill
- EP20 (step 356,320) abupt > 8.0% → kill
- EP25 (step 445,400) abupt > 6.9549% (baseline) → kill

## Baseline (target to beat)

| Metric | PR #174 (vu4jsiic, ep~45.3) | Target gap |
|--------|-----------------------------|-----------|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **6.9549** | < 6.7% to merge |
| `val_primary/wall_shear_y_rel_l2_pct` | 8.7345 | < 7.5% expected if hypothesis holds |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.5766 | < 9.0% expected if hypothesis holds |
| `val_primary/wall_shear_x_rel_l2_pct` | 7.109 | < 7.0% expected (small gain — dominant axis already aligned to flow) |

## Why this hypothesis is high-leverage

- AB-UPT achieves wsy=3.65%, wsz=3.63% — they almost certainly use a frame-aware shear representation; world-frame regression is the most obvious thing this codebase does that AB-UPT does not.
- Mirror-augmentation (tanjiro #443) flips y but cannot fix the underlying world-frame coupling; tangent-frame loss removes the coupling directly.
- Per-axis reweighting (haku #507) shifts gradient mass toward wsy/wsz but cannot solve a representation mismatch.
- Cost: zero extra params, ~5% extra compute for the rotation. If it works, it's a pure win.

## Constraints (do not break)

- `--no-compile-model` mandatory.
- `--fourier-pe` mandatory.
- `--no-use-ema` mandatory.
- Val/test metrics must remain in world frame for AB-UPT comparability.
- One PR, one hypothesis. No bundling other changes.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT gilbert: Your PR has been assigned. Please post your W&B run ID and confirm you have started training. The baseline to beat is **val_abupt = 6.9549%** (PR #174, run `vu4jsiic`). Post your ep1 metrics as soon as they are available and apply the standard kill gates (ep5: >13%, ep10: >10.5%, ep15: >8.5%, ep20: >8.0%).

**Comment by morganmcg1** [comment]
ADVISOR (infra-block escalation, 2026-05-03): Pod `senpai-bengio-gilbert` is confirmed blocked by a zombie `train.py` process from a prior wave. This is a known infrastructure issue (#466) — **you are not at fault**. The human infrastructure team has been notified and must SIGKILL the zombie process and restart the deployment.

Once the pod restarts and is free, launch your tangent-frame wall-shear loss run immediately per the PR instructions. Post your W&B run ID(s) and ep1 metrics as soon as available. No further action needed from you until the pod is freed.
